rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+:---------------|:-----------------------------------------------------|
+| **Author(s)** | Rick Chao, Tom O'Malley, Zhenyu Tan, Yuefeng Zhou (Google)                                  |
+| **Sponsor**   | Francois Chollet, Priya Gupta (Google)                                |
+| **Updated**   | 2020-11-21                                        |
+
+
+## Background
+
+With the recent release of TF2 parameter server training support ([ddoc](https://github.com/tensorflow/community/blob/master/rfcs/20200306-single-client-parameter-server.md)) ([API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/parameter_server_strategy_v2.py)) ([tutorial](https://www.tensorflow.org/tutorials/distribute/parameter_server_training)), custom training loop (CTL) users have started using the `ParameterServerStrategy` and `ClusterCoordinator` APIs for parameter server style distributed training. `ParameterServerStrategy` provides implementation of variable placement, and APIs for defining computation, and `ClusterCoordinator` provides APIs for dataset creation, asynchronous function scheduling and remote execution. The asynchronicity brought by `ClusterCoordinator` provides scalability and training fault tolerance, and at the same time implications such as the need for remote resource creation.
+
+While CTL user flow has since been supported, Keras `model.fit` training API is not yet. It has been a common ask (as shown in a survey conducted earlier this year) for availability, given its simplicity and support for a variety of machine learning models, metrics, optimizers, etc. 
+
+In this design, we will discuss the changes in `model.fit` API that we expect to make to accommodate asynchronous, coordinator-based parameter server training flow, and challenges the integration may have given the historical focus of synchronous distributed training with `model.fit`.
+
+
+## Goals
+
+
+
+*   Parameter server training support for Keras compile/fit style training API
+*   Minimal code changes across usage with other strategies
+*   Minimal performance implications
+
+
+## Proposed options and solutions
+
+Let’s first take a look at the proposed user flow (on the coordinator). It is expected to be largely the same with other strategies (except for the strategy swap). Unless mentioned otherwise, the discussion here applies to the python program intended to be run on the coordinator.
+
+
+### User Journey
+
+With a dataset factory:
+
+
+```
+cluster_resolver = ...
+strategy = tf.distribute.experimental.ParameterServerStrategy(cluster_resolver)
+with strategy.scope():
+  preproc_stage = ... # Some Keras preproc layers
+  model = ... # Building a Keras model
+model.compile(optimizer=..., loss=...)  # `ClusterCoordinator` is created
+def dataset_fn(): 
+  ... # Make use of `preproc_stage` for transformation
+history = model.fit(dataset_fn, epochs=..., steps_per_epoch=...,  callbacks=[...])
```

Comment:
Since the `ClusterCoordinator.schedule` can schedule a function to any worker. So it would be less error-prone to just ask users to create the same repeated, but shuffled differently, datasets on different workers. In this case, `steps_per_epoch` is needed.

It is right now difficult to achieve visitation guarantee with `ClusterCoordinator.schedule` but may be easier with `tf.data` service in the future.

Reply:
Does this mea we should *require* the user to return an infinite dataset in the PS case?

Reply:
Yes, we have explicitly documented the requirement: https://www.tensorflow.org/tutorials/distribute/parameter_server_training#more_about_dataset_creation

There is not error checking in `ParameterServerStrategy` now, but would be nice to add it.

Reply:
With the current design of parameter server training APIs, error handling allows OutOfRangeError to be reported to user, but the caveat is that it can mean only one worker has exhausted the dataset, while others still have some remaining to be processed. The behavior may likely be different from what the users would expect, especially when dataset is sharded, so the recommendation is to provide an infinite dataset that is shuffled with a different seed.

Reply:
SGTM, IMO then in the DataAdapter class for the DatasetFactory, we should add an error if the dataset isn't infinite that explains this

Reply:
Yes - added this in DataAdapter class part.